src/monomorphize.rs

```diff
@@ -11,3 +11,12 @@ pub fn apply_param_substs<'a, 'tcx,T>(tcx: &TyCtxt<'a, 'tcx, 'tcx>,
     let substituted = value.subst(*tcx, param_substs);
     tcx.normalize_associated_type(&substituted)
 }
+
+pub fn apply_ty_substs<'a, 'tcx>(tcx: &TyCtxt<'a, 'tcx, 'tcx>,
```

Comment:
Definitely we can share monomorphization with rustc at some point. rustc is currently in the process of overhauling how monomorphization works and the new scheme is much saner (you collect the set of monomorphizations required before trans instead of running trans recursively).

cc @eddyb @solson